airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/java/io/airbyte/cdk/integrations/destination/s3/S3StorageOperations.java

```diff
@@ -215,6 +217,45 @@ private String loadDataIntoBucket(final String objectPath, final SerializableBuf
     return newFilename;
   }
 
+  /**
+   * Users want deterministic file names (e.g. the first file part is really foo-0.csv Using UUIDs
+   * (previous approach) doesn't allow that. However, using pure integers could lead to a collision
+   * with an existing file. So, we'll count up the existing files in the directory and use that as a
+   * lazy-offset, assuming airbyte manages the dir and has similar naming conventions. `getPartId`
+   * will be 1-indexed.
+   */
+  @VisibleForTesting
+  String getPartId(String objectPath) {
+    if (partCounts.get(objectPath) == null) {
+      partCounts.put(objectPath, new AtomicInteger(0));
+    }
+
+    final AtomicInteger partCount = partCounts.get(objectPath);
```

Comment:
this entire if statement + subsequent assignment should be a single method call
```suggestion
    final AtomicInteger partCount = partCounts.computeIfAbsent(objectPath, p -> AtomicInteger::new);
```

computeIfAbsent is an atomic operation, whereas this if+put is vulnerable to a race condition (two threads both see that the map doesn't contain objectKey and `put`+`get` different objects)

Reply:
Nice!
 
I was justifying this with "sure, it'll multi-set, but the value is always 0 the firs time... so... 🤷" - That's some hard-life async node.js reasoning right there.

Reply:
resolved by 06c6aa439dfb3225b5cfeb60365c8484a36d9fb6